### PR TITLE
Unicode string test failure output

### PR DESCRIPTION
**Pull request by drathier**
This adds a unicode-escaped copy of any failing string that contains non-standard ascii characters. 

I'm not completely happy with the formatting, so if you have suggestions, please do tell. Otherwise, I think it's still much better than what we currently have. I'm comparing the escaped strings, not character by character, which is why we're getting diffs that cover part of an escape sequence. 

Solves https://github.com/elm-community/elm-test/issues/225 and #38.

```
↓ Expectations
↓ Expect.equal on unicode strings should show pretty output
✗ ascii
           ▼        ▼▼       ▼▼  
    "\u{1f63b}\u{1f640}\u{1f47b}" (same string but with unicode characters escaped)
     ▼   
    "😻🙀👻"
    |> Expect.equal
    "🙀👻😻🙈"
       ▲▲ 
    "\u{1f640}\u{1f47b}\u{1f63b}\u{1f648}" (same string but with unicode characters escaped)
           ▲▲▲▲▲▲▲▲▲▲        ▲▲      ▲ ▲  


↓ Expectations
↓ Expect.equal on unicode strings should show pretty output
✗ ascii space vs nbsp
        ▼    
    "asd qwe"
    ╵
    │ |> Expect.equal
    ╷
    "asd qwe"
        ▲    
    "asd\u{00a0}qwe" (same string but with unicode characters escaped)
        ▲▲▲▲▲▲    


↓ Expectations
↓ Expect.equal on unicode strings should show pretty output
✗ ascii only
         ▼   
    "asd qwe"
    ╵
    │ |> Expect.equal
    ╷
    "asd dwe"
         ▲   


↓ Expectations
↓ Expect.equal on unicode strings should show pretty output
✗ newline diff
                   
    "first\nsecond"
    ╵
    │ |> Expect.equal
    ╷
    "first\r\nsecond"
           ▲▲        


↓ Expectations
↓ Expect.equal on unicode strings should show pretty output
✗ long lines
                                                                                                                                                                                                                                                                                                                                                                                                            ▼▼▼▼▼▼▼▼▼▼                                                                                                                                                                                                                                                                                                     
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque in scelerisque arcu. Curabitur cursus efficitur turpis sed porttitor. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Nunc eu cursus ex. Proin accumsan quam quis dui semper finibus. Nunc vel nibh at tellus finibus rhoncus eu eget dolor. Sed eget neque ut lorem imperdiet fermentum. \u{1f63b} Morbi iaculis ante euismod, vulputate velit ut, varius velit. Nulla tempus dapibus mattis. In tempus, nisi a porta lobortis, nulla lacus iaculis quam, vel euismod magna risus at tortor. Integer porta urna odio. Nulla pellentesque dictum maximus. Donec auctor urna nec tortor imperdiet varius." (same string but with unicode characters escaped)
                                                                                                                                                                                                                                                                                                                                                                                                            ▼▼                                                                                                                                                                                                                                                                                                     
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque in scelerisque arcu. Curabitur cursus efficitur turpis sed porttitor. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Nunc eu cursus ex. Proin accumsan quam quis dui semper finibus. Nunc vel nibh at tellus finibus rhoncus eu eget dolor. Sed eget neque ut lorem imperdiet fermentum. 😻 Morbi iaculis ante euismod, vulputate velit ut, varius velit. Nulla tempus dapibus mattis. In tempus, nisi a porta lobortis, nulla lacus iaculis quam, vel euismod magna risus at tortor. Integer porta urna odio. Nulla pellentesque dictum maximus. Donec auctor urna nec tortor imperdiet varius."
    ╵
    │ |> Expect.equal
    ╷
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque in scelerisque arcu. Curabitur cursus efficitur turpis sed porttitor. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Nunc eu cursus ex. Proin accumsan quam quis dui semper finibus. Nunc vel nibh at tellus finibus rhoncus eu eget dolor. Sed eget neque ut lorem imperdiet fermentum. Morbi iaculis ante euismod, vulputate velit ut, varius velit. Nulla tempus dapibus mattis. In tempus, nisi a porta lobortis, nulla lacus iaculis quam, vel euismod magna risus at tortor. Integer porta urna odio. Nulla pellentesque dictum maximus. Donec auctor urna nec tortor imperdiet varius."
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 


```

### CODE REVIEW

**Comment by drathier** [comment]
We seem to have multiple copies of `verticalBar` etc., and this only updates the ones used in `npm test`. I'm wondering a bit why we have multiple copies of them?

**Comment by harrysarson** [comment]
How does it look if the strings are long or contain new lines?

**Comment by drathier** [comment]
Added a newline test (\n vs \r\n), those are typed out as their backslash escapes (this is how elm types them out in toString). Also added a very long line, hard to read test output in my terminal because of word wrap, but very easy without word wrap.

File: tests/src/Runner/String/Format.elm [hunk 1]
```diff
@@ -66,16 +66,84 @@ format description reason =
...
+                                    "d"
+
+                                14 ->
+                                    "e"
+
+                                15 ->
+                                    "f"
+
+                                _ ->
+                                    String.fromInt (i |> remainderBy 16)
```
Comment (mgold) [inline tests/src/Runner/String/Format.elm]: No need to call `remainderBy` again. `decimalDigit -> String.fromInt decimalDigit`

File: tests/src/Runner/String/Format.elm [hunk 2]
```diff
@@ -66,16 +66,84 @@ format description reason =
...
+                && (-- except ` and ^ because they're combining marks, like in è and ê.
+                    List.member v [ 94, 96 ] == False
+                   )
+    in
+    s
+        |> String.toList
+        |> List.map Char.toCode
+        |> List.map
+            (\c ->
+                if isAsciiChar c == False then
```
Comment (mgold) [inline tests/src/Runner/String/Format.elm]: Use `if not (isAsciiChar c) then` or flip the conditional.

**Comment by drathier** [comment]
Better late than never, I guess.

File: tests/src/Tests.elm [hunk 3]
```diff
@@ -83,6 +83,19 @@ expectationTests =
...
+        --     ]
```
Comment (harrysarson) [inline tests/src/Tests.elm]: Why are these tests commented out?

File: tests/src/Runner/String/Format.elm [hunk 4]
```diff
@@ -135,29 +203,64 @@ equalityToString : { operation : String, expected : String, actual : String } ->
 equalityToString { operation, expected, actual } =
     -- TODO make sure this looks reasonable for multiline strings
```
Comment (harrysarson) [inline tests/src/Runner/String/Format.elm]: I think this pull request solves this TODO. 🙂

File: tests/src/Tests.elm (hunk 3 above)
Comment (drathier) [reply tests/src/Tests.elm]: They're the tests used to show the output of failed string equality expectations. I don't know how to write meaningful automatic tests that check that the output remains the same.

File: tests/src/Runner/String/Format.elm (hunk 4 above)
Comment (drathier) [reply tests/src/Runner/String/Format.elm]: Unfortunately not :/ Long strings with automatic line wrapping still look weird.

**Pull request merged**